In [2]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

In [3]:
# Load the Iris dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=37)


In [4]:
# Standardize the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [5]:
# Add a bias term to the feature matrix
X_train_biased = np.c_[np.ones((X_train_scaled.shape[0], 1)), X_train_scaled]
X_test_biased = np.c_[np.ones((X_test_scaled.shape[0], 1)), X_test_scaled]


In [6]:
# One-hot encode the target variable
num_classes = len(np.unique(y))
y_one_hot = np.eye(num_classes)[y_train]

In [7]:
# Softmax function
def softmax(x):
    exp_x = np.exp(x - np.max(x, axis=1, keepdims=True))
    return exp_x / np.sum(exp_x, axis=1, keepdims=True)


In [8]:
# Initialize weights
num_features = X_train_biased.shape[1]
num_classes = len(np.unique(y))
weights = np.random.randn(num_features, num_classes)

# Training parameters
learning_rate = 0.01
num_epochs = 1000


for epoch in range(num_epochs):
    # Forward pass
    logits = np.dot(X_train_biased, weights)
    probabilities = softmax(logits)

    # Compute loss (cross-entropy)
    loss = -np.sum(y_one_hot * np.log(probabilities)) / len(X_train_biased)

    # Compute gradients
    gradients = np.dot(X_train_biased.T, (probabilities - y_one_hot)) / len(X_train_biased)

    # Update weights using gradient descent
    weights -= learning_rate * gradients
   # Print loss for every 100 epochs
    if epoch % 100 == 0:
        print(f'Epoch {epoch}, Loss: {loss}')



Epoch 0, Loss: 0.7876198724878628
Epoch 100, Loss: 0.6495313807261618
Epoch 200, Loss: 0.5563416222733442
Epoch 300, Loss: 0.49312517410385515
Epoch 400, Loss: 0.4488597109920474
Epoch 500, Loss: 0.41652247747573556
Epoch 600, Loss: 0.39186391702277085
Epoch 700, Loss: 0.3723094468551976
Epoch 800, Loss: 0.3562606224660867
Epoch 900, Loss: 0.34269558410549966


In [9]:
 # Make predictions on the test set
test_logits = np.dot(X_test_biased, weights)
test_probabilities = softmax(test_logits)
y_pred = np.argmax(test_probabilities, axis=1)


In [10]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy on the test set: {accuracy}')

Accuracy on the test set: 0.9333333333333333


An accuracy of 0.9333 on the test set indicates that  softmax regression model is performing well on the Iris dataset. This accuracy suggests that the model is making correct predictions for approximately 93.33% of the instances in the test set.